crates/kitsune_p2p/fetch/src/pool.rs

```diff
@@ -458,25 +469,120 @@ mod tests {
 
-        assert_eq!(ss.next(sec1), Some(source(1)));
-        assert_eq!(ss.next(sec1), Some(source(2)));
-        assert_eq!(ss.next(sec1), None);
+        assert_eq!(sources.next(source_delay), Some(source(1)));
+        assert_eq!(sources.next(source_delay), Some(source(2)));
+        assert_eq!(sources.next(source_delay), None);
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn source_rotation_prioritises_less_recently_tried() {
+        let source_delay = Duration::from_secs(10);
+        let mut sources = Sources(vec![
+            SourceRecord {
+                source: source(1),
+                last_request: Some(Instant::now()), // recently tried
+            }
+            .into(),
+            SourceRecord {
+                source: source(2),
+                last_request: None, // never checked
+            }
+            .into(),
+            SourceRecord {
+                source: source(3),
+                last_request: None, // never checked
+            }
+            .into(),
+        ]);
+
+        assert_eq!(sources.next(source_delay), Some(source(2)));
+
+        // All sources now past their retry delay
+        tokio::time::advance(source_delay).await;
+
+        // Would expect source 1 to be tried next but trying 2 first rotated 1 and 2 to the end of the list
+        assert_eq!(sources.next(source_delay), Some(source(3)));
+        assert_eq!(sources.next(source_delay), Some(source(1)));
+        assert_eq!(sources.next(source_delay), Some(source(2)));
+        assert_eq!(sources.next(source_delay), None);
+    }
+
+    #[tokio::test(start_paused = true)]
+    async fn source_rotation_uses_all_sources() {
+        let mut noise = [0; 1_000];
+        rand::thread_rng().fill_bytes(&mut noise);
+        let mut u = Unstructured::new(&noise);
+
+        let source_delay = Duration::from_secs(rand::thread_rng().gen_range(1..50));
+
+        let v = std::iter::repeat_with(|| Duration::arbitrary(&mut u).unwrap())
+            .take(100)
+            .enumerate()
+            .map(|(i, duration)| SourceRecord {
+                source: source(i as u8),
+                last_request: Instant::now().checked_sub(duration),
+            })
+            .collect();
+        let mut sources = Sources(v);
+
+        let mut seen_sources: HashSet<u8> = HashSet::new();
+        for _ in 0..100 {
+            if let Some(s) = sources.next(source_delay) {
+                match s {
+                    FetchSource::Agent(a) => {
+                        // The source agent key is a repeating byte array of the source number, so we can retrieve any
+                        // byte here to get the source number
+                        seen_sources.insert(a.0[0]);
+                    }
+                }
+            }
+
+            tokio::time::advance(source_delay).await;
+        }
+
+        assert_eq!(100, seen_sources.len());
+    }
+
+    #[test]
+    fn state_removes_context_on_merge_if_new_is_none() {
```

Comment:
We use the context as a flag that should only be settable, not un-settable, to let us know that the a validation receipt was requested. Even if they also happened to request the op through gossip, we still want to send the receipt.